### PR TITLE
Fix aten.logspace decomposition

### LINKED ISSUE #104118: [torch.compile] FakeTensorMode not working with torch.logspace on complex arguments
### 🐛 Describe the bug

my repro:
```
from torch import _dynamo as dy

def func():
    a = torch.logspace(3, 10 + 3j, steps=5)


func()
opt_func = dy.optimize('eager')(func)
opt_func()
```

stacktrace:
```
Traceback (most recent call last):
  File "examples/wip.py", line 19, in <module>
    opt_func()
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 286, in _fn
    return fn(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 439, in catch_errors
    return callback(frame, cache_size, hooks, frame_state)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 522, in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 125, in _fn
    return fn(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 358, in _convert_frame_assert
    return _compile(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 177, in time_wrapper
    r = func(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 428, in _compile
    out_code = transform_code_object(code, transform)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/bytecode_transformation.py", line 1000, in transform_code_object
    transformations(instructions, code_options)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/convert_frame.py", line 413, in transform
    tracer.run()
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 2009, in run
    super().run()
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 703, in run
    and self.step()
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 663, in step
    getattr(self, inst.opname)(inst)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 385, in wrapper
    return inner_fn(self, inst)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 1146, in CALL_FUNCTION_KW
    self.call_function(fn, args, kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/symbolic_convert.py", line 554, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/variables/torch.py", line 565, in call_function
    tensor_variable = wrap_fx_proxy(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 1003, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/variables/builder.py", line 1038, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 1288, in get_fake_value
    raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 1256, in get_fake_value
    return wrap_fake_exception(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 850, in wrap_fake_exception
    return fn()
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 1257, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 1322, in run_node
    raise RuntimeError(fn_str + str(e)).with_traceback(e.__traceback__) from e
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 1309, in run_node
    return node.target(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 1105, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 1306, in dispatch
    op_impl_out = op_impl(self, func, *args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_subclasses/fake_tensor.py", line 385, in constructors
    r = func(*args, **new_kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_ops.py", line 398, in __call__
    return self._op(*args, **kwargs or {})
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_prims_common/wrappers.py", line 227, in _fn
    result = fn(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_refs/__init__.py", line 4567, in logspace
    ret = torch.linspace(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_prims_common/wrappers.py", line 227, in _fn
    result = fn(*args, **kwargs)
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_refs/__init__.py", line 4485, in linspace
    check(
  File "/home/yshi/miniconda3/envs/dy/lib/python3.8/site-packages/torch/_prims_common/__init__.py", line 1648, in check
    raise exc_type(s())
torch._dynamo.exc.TorchRuntimeError: Failed running call_function <built-in method logspace of type object at 0x7f6c1be42f60>(*(3, (10+3j)), **{'steps': 5}):
linspace(): inferred dtype torch.complex64 can't be safely cast to passed dtype torch.float64

from user code:
   File "examples/wip.py", line 13, in func
    a = torch.logspace(3, 10 + 3j, steps=5)

Set torch._dynamo.config.verbose=True or TORCHDYNAMO_VERBOSE=1 for more information


You can suppress this exception and fall back to eager by setting:
    import torch._dynamo
    torch._dynamo.config.suppress_errors = True
```

Note that non-complex arguments work fine

### Versions

PyTorch version: 2.1.0.dev20230516+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 10.4.0-4ubuntu1~22.04) 10.4.0
Clang version: 14.0.0-1ubuntu1
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.8.16 (default, Jun 12 2023, 18:09:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.0-45-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070
Nvidia driver version: 520.61.05
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 9 5900X 12-Core Processor
CPU family:                      25
Model:                           33
Thread(s) per core:              2
Core(s) per socket:              12
Socket(s):                       1
Stepping:                        0
Frequency boost:                 enabled
CPU max MHz:                     4950.1948
CPU min MHz:                     2200.0000
BogoMIPS:                        7399.95
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization:                  AMD-V
L1d cache:                       384 KiB (12 instances)
L1i cache:                       384 KiB (12 instances)
L2 cache:                        6 MiB (12 instances)
L3 cache:                        64 MiB (2 instances)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] numpy==1.24.1
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230516+cu118
[pip3] torchaudio==2.1.0.dev20230516+cu118
[pip3] torchvision==0.16.0.dev20230516+cu118
[conda] numpy                     1.24.1                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230516+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230516+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230516+cu118          pypi_0    pypi

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by yanboliang**
Fixes #104118

### CODE REVIEW

**Comment by soulitzer** [comment]
ci failures look real

**Review by ezyang — APPROVED** [review]

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -4619,20 +4619,29 @@ def logspace(
         if isinstance(end, FloatLike):
             end = sym_int(end)
 
+    if py_any(isinstance(arg, complex) for arg in (start, end, steps)):
+        default_complex_dtype = utils.corresponding_complex_dtype(
+            torch.get_default_dtype()
+        )
+        dtype = default_complex_dtype
+        _dtype = None  # torch.linspace will update the correct dtype
```
Comment (soulitzer) [inline torch/_refs/__init__.py:4627]: Doesn't linspace also make it the `default_complex_dtype`? Is there any difference between passing in None and passing in dtype?

File: test/test_prims.py [hunk 2]
```diff
@@ -1212,6 +1212,16 @@ def test_unbind(self):
...
 
+    def test_logspace_with_complex_input(self):
+        actual = refs.logspace(2, 10 + 5j, steps=5)
+        expect = torch.logspace(2, 10 + 5j, steps=5)
+        self.assertEqual(actual, expect)
+
+    def test_linspace_with_complex_input(self):
+        actual = refs.linspace(2, 10 + 5j, steps=5)
+        expect = torch.linspace(2, 10 + 5j, steps=5)
+        self.assertEqual(actual, expect)
```
Comment (soulitzer) [inline test/test_prims.py:1223]: Would be cool to test these within the OpInfo sample inputs instead of having one off tests

**Review by soulitzer — COMMENTED** [review]
Looks good, just minor comments

File: torch/_refs/__init__.py (hunk 1 above)
Comment (yanboliang) [reply torch/_refs/__init__.py:4627]: They are equivalent, we can do either way.

**Pull request closed without merging**